Src/Compilers/Core/MSBuildTask/BuildClient.cs

```diff
@@ -0,0 +1,398 @@
+            Process[] allProcesses = Process.GetProcessesByName(PipeName);
+            Log("Found {0} existing processes with matching base name", allProcesses.Length);
+
+            foreach (Process process in allProcesses)
+            {
+                using (process)
+                {
+                    int processId = 0;
+                    try
+                    {
+                        var exeNameBuffer = new StringBuilder(MAX_PATH_SIZE);
+                        int pathSize = MAX_PATH_SIZE;
+                        string fullFileName = null;
+
+                        if (QueryFullProcessImageName(process.Handle,
+                                                      0, // Win32 path format
+                                                      exeNameBuffer,
+                                                      ref pathSize))
+                        {
+                            fullFileName = exeNameBuffer.ToString();
+                            Log("Process file path: {0}", fullFileName);
+                        }
+
+                        if (string.Equals(fullFileName,
+                                          expectedPath,
+                                          StringComparison.OrdinalIgnoreCase))
+                        {
+                            processId = process.Id;
+                        }
+                    }
+                    // If any exception occurs (e.g., accessing the process handle
+                    // fails because the process is exiting), we should simply fail
+                    // to connect and let the loop fall through
+                    catch
+                    { }
+
+                    if (processId != 0)
+                        yield return processId;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Connect to the given process id and return a pipe.
+        /// Throws on cancellation.
+        /// </summary>
+        /// <param name="processId">Process id to try to connect to.</param>
+        /// <param name="timeoutMs">Timeout to allow in connecting to process.</param>
+        /// <param name="cancellationToken">Cancellation token to cancel connection to server.</param>
+        /// <returns>
+        /// An open <see cref="NamedPipeClientStream"/> to the server process or null on failure.
+        /// </returns>
+        private static NamedPipeClientStream TryConnectToProcess(
+            int processId,
+            int timeoutMs,
+            CancellationToken cancellationToken)
+        {
+            NamedPipeClientStream pipeStream;
+            try
+            {
+                // Machine-local named pipes are named "\\.\pipe\<pipename>".
+                // We use the pipe name followed by the process id.
+                // The NamedPipeClientStream class handles the "\\.\pipe\" part for us.
+                string pipeName = PipeName + processId.ToString();
```

Comment:
ToString() is unnecessary here (even if you care about boxing). The compiler currently avoid boxing when concatenating primitive value type to a string.